File: backend/app/Services/InviteService.php
<?php

namespace App\Services;

use App\Models\Friend;
use App\Models\Invite;
use App\Models\User;
use Carbon\Carbon;
use Illuminate\Support\Str;

/**
 * Personal-invite business logic (Feature 5).
 *
 * The single seam for minting, resolving, and connecting via invite codes,
 * kept provider-agnostic so a paid deferred-deep-link provider (Branch / Adjust
 * / …) can be added later as configuration, not surgery (DECISION D4): callers
 * only ever deal in the opaque code, never a provider SDK.
 */
class InviteService
{
    /**
     * Return the caller's reusable invite code, minting one on first use.
     *
     * One code per inviter (the `invites.inviter_id` unique constraint), so
     * repeated calls are idempotent and the shared link is stable.
     */
    public function mintFor(User $inviter): Invite
    {
        $existing = Invite::firstWhere('inviter_id', $inviter->id);
        if ($existing !== null) {
            return $existing;
        }

        return Invite::create([
            'inviter_id' => $inviter->id,
            'code' => $this->uniqueCode(),
        ]);
    }

    /**
     * Resolve a code to its invite (with the inviter eager-loaded), or null.
     */
    public function resolve(string $code): ?Invite
    {
        return Invite::with('inviter')->firstWhere('code', $code);
    }

    /**
     * Records a step of the invite loop against [$code].
     *
     * The loop is: link shared -> landing page opened -> web demo watched ->
     * store tapped -> app installed -> account connected. Everything between
     * "shared" and "connected" was invisible, so there was no way to tell
     * whether the web demo earned its place or where the loop leaked.
     *
     * Counted, not flagged: a link dropped into a group chat is opened many
     * times, and that reach is the thing worth knowing.
     *
     * Silent when the code is unknown. This is reachable from a public page, so
     * a bad code is a typo or a probe, not an error worth reporting.
     *
     * @param  string  $code  The invite code from the URL.
     * @param  string  $step  One of `opened`, `demo_completed`, `store_clicked`.
     */
    public function recordFunnelStep(string $code, string $step): void
    {
        $column = match ($step) {
            'opened' => 'opened_count',
            'demo_completed' => 'demo_completed_count',
            'store_clicked' => 'store_clicked_count',
            default => null,
        };
        if ($column === null) {
            return;
        }

        $invite = Invite::firstWhere('code', $code);
        if (! $invite) {
            return;
        }

        // increment() is a single atomic UPDATE, so two people opening the same
        // link at once cannot lose a count to a read-modify-write race.
        $invite->increment($column);

        if ($step === 'opened' && $invite->first_opened_at === null) {
            $invite->forceFill(['first_opened_at' => now()])->saveQuietly();
        }
    }

    /**
     * Connect an arriving invitee to the inviter: create the mutual friendship
     * (idempotent). A self-referential code is a no-op. Reuses the same
     * two-row friendship shape as accepting a friend request.
     */
    public function connect(Invite $invite, User $invitee): void
    {
        if ($invite->inviter_id === $invitee->id) {
            return;
        }

        $this->linkFriends($invite->inviter_id, $invitee->id);
        $this->linkFriends($invitee->id, $invite->inviter_id);
    }

    /** Insert one directed friendship row if it doesn't already exist. */
    private function linkFriends(int $userId, int $friendId): void
    {
        Friend::firstOrCreate(
            ['user_id' => $userId, 'friend_id' => $friendId],
            ['became_friends_at' => Carbon::now()],
        );
    }

    /** Generate an opaque, URL-safe code not already in use. */
    private function uniqueCode(): string
    {
        do {
            $code = Str::lower(Str::random(12));
        } while (Invite::where('code', $code)->exists());

        return $code;
    }
}

File: app/lib/features/invite/data/invite_service.dart
import '../../../networks/dio/dio.dart';
import '../../../networks/endpoints.dart';
import '../model/inviter.dart';

/// Client seam for personal invites (Feature 5).
///
/// Kept provider-agnostic — callers only ever deal in the opaque code and the
/// `reacti.app/i/{code}` link — so a paid deferred-deep-link provider (Branch /
/// Adjust / …) can be added later as configuration, not surgery (DECISION D4).
///
/// The mutable [instance] is the seam tests swap with a fake.
class InviteService {
  /// The shared instance (swappable in tests).
  static InviteService instance = InviteService();

  /// The shareable link for [code], on the same host as the API (so a staging
  /// build links to staging.reacti.io and prod to reacti.io) — never the dead
  /// reacti.app. The backend serves a landing page at `/i/{code}`.
  String linkFor(String code) {
    final host = url.replaceFirst(RegExp(r'/api/?$'), '');
    return '$host/i/$code';
  }

  /// Mint (or fetch) the caller's reusable invite code; null on failure.
  Future<String?> mintCode() async {
    try {
      final res = await postHttp(EndPoints.mintInvite());
      if (res.statusCode == 200) {
        return (res.data['data']?['code']) as String?;
      }
    } catch (_) {
      // Swallow — the caller shows a soft failure, never blocks the user.
    }
    return null;
  }

  /// Resolve [code] to the inviter's public profile; null when unknown.
  Future<Inviter?> resolveInviter(String code) async {
    try {
      final res = await getHttp(EndPoints.resolveInvite(code));
      final data = res.data['data'];
      if (res.statusCode == 200 && data is Map) {
        return Inviter.fromJson(Map<String, dynamic>.from(data));
      }
    } catch (_) {}
    return null;
  }

  /// Connect (befriend the inviter) via [code]; returns the inviter id or null.
  Future<int?> connect(String code) async {
    try {
      final res = await postHttp(EndPoints.connectInvite(code));
      if (res.statusCode == 200) {
        return (res.data['data']?['inviter_id']) as int?;
      }
    } catch (_) {}
    return null;
  }

  /// Extracts an invite code from pasted [input]: a `reacti.app/i/{code}` link
  /// (or any URL containing `/i/{code}`) or a bare alphanumeric code. Null when
  /// nothing usable is found.
  String? codeFromInput(String input) {
    final trimmed = input.trim();
    if (trimmed.isEmpty) return null;

    final inLink = RegExp(r'/i/([A-Za-z0-9]+)').firstMatch(trimmed);
    if (inLink != null) return inLink.group(1);

    if (RegExp(r'^[A-Za-z0-9]+$').hasMatch(trimmed)) return trimmed;
    return null;
  }
}
